console/console-init/ui/src/modules/device/dialogs/CreateDevice/MetaDataRow.tsx

```diff
@@ -11,108 +11,172 @@ import {
+  // console.log("metadataList", metadataList, "type", metadataRow.type);
   const [propertySelected, setPropertySelected] = useState<string | undefined>(
     ""
   );
-  const [propertyInput, setPropertyInput] = useState<string | undefined>("");
+  const [isExpanded, setIsExpanded] = useState<boolean>(false);
+
+  const onToggle = (isExpanded: boolean) => {
+    setIsExpanded(isExpanded);
+  };
+
+  const onTypeAheadSelect = (e: any, selection: SelectOptionObject) => {
+    onPropertySelect && onPropertySelect(e, selection);
+    setIsExpanded(false);
+  };
+
+  const onSelectType = (typeValue: string) => {
+    let updatedTypeMetadata = [...metadataList];
+    updatedTypeMetadata[rowIndex].type = typeValue;
+    setMetadataList(updatedTypeMetadata);
+  };
+
+  //TODO: Move this method to container
+  const onFilter = (e: any) => {
+    const propertyInput = e.target.value && e.target.value.trim();
+    //TODO: Analyze if option type can be changed
+    let propertyOptions: React.ReactElement[] = [];
+    onChangePropertyInput &&
+      onChangePropertyInput(propertyInput).then((options: any) => {
+        const propertyList = options;
+        propertyOptions = propertyList
+          ? propertyList.map((propertyItem: any, index: number) => (
+              <SelectOption
+                disabled={propertyItem.isDisabled}
+                key={index}
+                value={propertyItem.value}
+              />
+            ))
+          : [];
+      });
+
+    return propertyOptions;
+  };
+
+  //TODO: Call graphql queries and populate options in SelectOption
+  const onChangePropertyInput = async (propertyKey: string) => {
+    let updatedPropertyMetadata = [...metadataList];
+    updatedPropertyMetadata[rowIndex].key = propertyKey;
+    setMetadataList(updatedPropertyMetadata);
+  };
 
   const handleAddChildRow = (event: any, parentKey: string) => {
-    let element = { defaults: { parentKey: {} }, ext: {} };
-    setFormData([...formData, element]);
+    let newRow = {
+      key: parentKey + "/",
+      value: [],
+      type: deviceRegistrationTypeOptions[0].value
+    };
+    setMetadataList([...metadataList, newRow]);
   };
 
   const handleDeleteRow = (index: any) => {
-    const metadataList = [...formData];
-    metadataList.splice(index, 1);
-    setFormData(metadataList);
+    const deletedRowMetadata = [...metadataList];
+    deletedRowMetadata.splice(index, 1);
+    setMetadataList(deletedRowMetadata);
   };
 
-  const handleInputChange = (e: any, index: any) => {
-    const { name, value } = e.target;
-    let metadataList = [...formData];
-    // metadataList[index][name]=value;
-    setFormData(metadataList);
+  const handleValueChange = (propertyValue: string, e: any) => {
+    let updatedValueMetadata = [...metadataList];
+    updatedValueMetadata[rowIndex].value = propertyValue;
+    setMetadataList(updatedValueMetadata);
   };
 
   const onPropertySelect = (e: any, selection: SelectOptionObject) => {
```

Comment:
I think we can have function/event name onSelectProperty, onClearProperty. event name should come first then postfix. 
onSelect, onClear, onChange etc. wdyt?

Reply:
Agreed.